src/components/verification-grid/verification-grid.ts

```diff
@@ -0,0 +1,1364 @@
+      const newTargetUrl = item.url;
+
+      // we can't directly compare the two objects because their strict equality
+      // will always fail because it will be comparing two different memory
+      // addresses. To prevent each tile re-rendering, we can selectively
+      // re-render only he tiles which have changed sources
+      if (currentTargetUrl !== newTargetUrl) {
+        tile.model = item;
+        tile.index = i;
+      }
+    });
+
+    // if we are on the last page, we hide the remaining elements
+    const pagedDelta = elements.length - nextPage.length;
+    if (pagedDelta > 0) {
+      this.hideGridItems(pagedDelta);
+    } else if (this.hiddenTiles > 0) {
+      this.showAllGridItems();
+      this.hiddenTiles = 0;
+    }
+
+    this.cacheNext();
+  }
+
+  private hideGridItems(numberOfTiles: number): void {
+    Array.from(this.gridTiles)
+      .slice(-numberOfTiles)
+      .forEach((element) => {
+        element.hidden = true;
+      });
+
+    this.hiddenTiles = numberOfTiles;
+  }
+
+  private showAllGridItems(): void {
+    const gridTiles = this.gridTiles ?? [];
+    for (const element of gridTiles) {
+      element.hidden = false;
+    }
+  }
+
+  private createSpectrogramElements(): void {
+    // we use a buffer so that the entire component doesn't re-render
+    // every time that we add a spectrogram element
+    const verificationGridBuffer = [];
+
+    // TODO: we might be able to do partial rendering or removal if some of the
+    // TODO: we might be able to set the OE verification model when creating the elements
+    // needed spectrogram elements already exist
+    for (let i = 0; i < this.gridSize; i++) {
+      const template = this.gridItemTemplate.content.cloneNode(true);
+      verificationGridBuffer.push(
+        html`<oe-verification-grid-tile @loaded="${this.handleSpectrogramLoaded}">
+          ${template}
+        </oe-verification-grid-tile>`,
+      );
+    }
+
+    this.spectrogramElements = verificationGridBuffer;
+
+    // we store the grid tiles inside a const variable so that we don't have
+    // to do a DOM query in every loop iteration
+    const gridTiles = this.gridTiles ?? [];
+    for (const element of gridTiles) {
+      this.intersectionObserver.observe(element);
+    }
+  }
+
+  private areSpectrogramsLoaded(): boolean {
+    const gridTilesArray = Array.from(this.gridTiles);
+    return !gridTilesArray.some((tile: VerificationGridTileComponent) => !tile.loaded);
+  }
+
+  private handleSpectrogramLoaded(): void {
+    const decisionsDisabled = this.areDecisionsDisabled();
+    const loading = !this.areSpectrogramsLoaded();
+
+    if (decisionsDisabled !== loading) {
+      this.setDecisionDisabled(loading);
+    }
+
+    if (!loading) {
+      this.dispatchEvent(new CustomEvent(VerificationGridComponent.loadedEventName));
+    }
+  }
+
+  //#endregion
+
+  //#region Caching
+
+  private async cacheClient(elapsedItems: number): Promise<void> {
```

Comment:
good candidate for a service class

Reply:
Created `VerificationGridFetcher` which handles fetching resources and implements our caching strategy.
It also implements our new paging strategy of having a `queueBuffer` so that we no longer need to a store `pagedItems` count.

The new verification grid page fetcher fetches a full page of `Verification` models, what is not used from the paging callback is put into a queueBuffer.
I specify page callback results, because I'm not storing audio files in a queueBuffer, only the `Verification` models that we use that has a pointer to the audio file).

When a subsequent request comes in for items, we try to use the callback queue items before using the paging callback again.
Meaning that if we do a partial page, we can pull straight from the queue without using the `getPage` callback again to fetch more `Verification` models (which may involve API calls).

This approach also improves performance for sub-selection verifications.